File: Sources/KakaoSapiens/Services/SyncDeviceListModel.swift
import Combine
import Foundation

public struct SyncAccountDevice: Equatable, Identifiable {
    public let id: String
    public let platform: String
    public let linkedAt: String
    public let isCurrent: Bool

    public var title: String {
        switch platform {
        case "macos": return "Mac"
        case "android_phone": return "Android 폰"
        case "android_tablet": return "Android 태블릿"
        default: return "알 수 없는 기기"
        }
    }
}

public enum SyncDeviceListState: Equatable {
    case idle
    case loading
    case loaded([SyncAccountDevice])
    case failed
}

@MainActor
public final class SyncDeviceListModel: ObservableObject {
    @Published public private(set) var state: SyncDeviceListState = .idle
    /// The device the user has asked to remove, waiting for confirmation.
    @Published public private(set) var pendingRevoke: SyncAccountDevice?
    @Published public private(set) var revokeFailed = false
    private let client: SyncWorkerClient

    public init(client: SyncWorkerClient) { self.client = client }

    /// Removing *this* device is deliberately not offered here.
    ///
    /// Revoking is a server-side act; it would leave this Mac holding keys the
    /// account no longer honours — the half-linked state this whole feature
    /// exists to escape. Leaving is what "이 기기 연결 해제" does, and that
    /// clears the local side too.
    public func canRevoke(_ device: SyncAccountDevice) -> Bool { !device.isCurrent }

    public func requestRevoke(_ device: SyncAccountDevice) {
        guard canRevoke(device) else { return }
        revokeFailed = false
        pendingRevoke = device
    }

    public func cancelRevoke() {
        pendingRevoke = nil
        revokeFailed = false
    }

    public func confirmRevoke() async {
        guard let device = pendingRevoke else { return }
        pendingRevoke = nil
        do {
            _ = try await client.revokeDevice(deviceID: device.id)
            revokeFailed = false
            // Read the list back rather than removing the row locally: what the
            // account holds is the server's answer, not our guess at it.
            state = .idle
            await load()
        } catch {
            revokeFailed = true
        }
    }

    public func load() async {
        guard state != .loading else { return }
        state = .loading
        do {
            let response = try await client.devices()
            state = .loaded(try Self.parse(response.body))
        } catch {
            state = .failed
        }
    }

    private static func parse(_ data: Data) throws -> [SyncAccountDevice] {
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let result = root["result"] as? [String: Any],
              let rows = result["devices"] as? [[String: Any]] else {
            throw SyncWorkerClientError.invalidResponse
        }
        return try rows.map { row in
            guard let id = row["device_id"] as? String,
                  let platform = row["platform"] as? String,
                  let linkedAt = row["linked_at"] as? String,
                  let isCurrent = row["is_current"] as? Bool,
                  !id.isEmpty,
                  ["macos", "android_phone", "android_tablet"].contains(platform),
                  ISO8601DateFormatter().date(from: linkedAt) != nil else {
                throw SyncWorkerClientError.invalidResponse
            }
            return SyncAccountDevice(id: id, platform: platform, linkedAt: linkedAt, isCurrent: isCurrent)
        }
    }
}

File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncDeviceListModel.kt
package com.sapiens.gagaodok.sync

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import java.time.Instant

data class SyncAccountDevice(
    val id: String,
    val platform: String,
    val linkedAt: String,
    val isCurrent: Boolean,
) {
    val title: String get() = when (platform) {
        "macos" -> "Mac"
        "android_phone" -> "Android 폰"
        "android_tablet" -> "Android 태블릿"
        else -> "알 수 없는 기기"
    }
}

sealed interface SyncDeviceListState {
    data object Idle : SyncDeviceListState
    data object Loading : SyncDeviceListState
    data class Loaded(val devices: List<SyncAccountDevice>) : SyncDeviceListState
    data object Failed : SyncDeviceListState
}

class SyncDeviceListModel(private val client: SyncWorkerClient) {
    private val mutableState = MutableStateFlow<SyncDeviceListState>(SyncDeviceListState.Idle)
    val state: StateFlow<SyncDeviceListState> = mutableState.asStateFlow()

    /** The device the user asked to remove, waiting for confirmation. */
    private val mutablePendingRevoke = MutableStateFlow<SyncAccountDevice?>(null)
    val pendingRevoke: StateFlow<SyncAccountDevice?> = mutablePendingRevoke.asStateFlow()
    private val mutableRevokeFailed = MutableStateFlow(false)
    val revokeFailed: StateFlow<Boolean> = mutableRevokeFailed.asStateFlow()

    /**
     * Removing *this* device is deliberately not offered here.
     *
     * Revoking is a server-side act; it would leave this device holding keys
     * the account no longer honours — the half-linked state this whole feature
     * exists to escape. Leaving is what "이 기기 연결 해제" does, and that
     * clears the local side too.
     */
    fun canRevoke(device: SyncAccountDevice): Boolean = !device.isCurrent

    fun requestRevoke(device: SyncAccountDevice) {
        if (!canRevoke(device)) return
        mutableRevokeFailed.value = false
        mutablePendingRevoke.value = device
    }

    fun cancelRevoke() {
        mutablePendingRevoke.value = null
        mutableRevokeFailed.value = false
    }

    fun confirmRevoke() {
        val device = mutablePendingRevoke.value ?: return
        mutablePendingRevoke.value = null
        runCatching { client.revokeDevice(device.id) }
            .onSuccess {
                mutableRevokeFailed.value = false
                // Read the list back rather than dropping the row locally:
                // what the account holds is the server's answer, not our guess.
                mutableState.value = SyncDeviceListState.Idle
                load()
            }
            .onFailure { mutableRevokeFailed.value = true }
    }

    fun load() {
        if (mutableState.value == SyncDeviceListState.Loading) return
        mutableState.value = SyncDeviceListState.Loading
        mutableState.value = runCatching { SyncDeviceListState.Loaded(parse(client.devices().body)) }
            .getOrElse { SyncDeviceListState.Failed }
    }

    private fun parse(body: ByteArray): List<SyncAccountDevice> {
        val rows = JSONObject(String(body, Charsets.UTF_8)).getJSONObject("result").getJSONArray("devices")
        return List(rows.length()) { index ->
            val row = rows.getJSONObject(index)
            val id = row.getString("device_id").also { require(it.isNotEmpty()) }
            val platform = row.getString("platform").also {
                require(it in setOf("macos", "android_phone", "android_tablet"))
            }
            val linkedAt = row.getString("linked_at").also { Instant.parse(it) }
            SyncAccountDevice(id, platform, linkedAt, row.getBoolean("is_current"))
        }
    }
}
